M/MPFR/build_tarballs.jl

```diff
@@ -34,7 +34,7 @@ products = [
 
 # Dependencies that must be installed before this package can be built
 dependencies = [
-    Dependency("GMP_jll", v"6.2.0"),
+    Dependency("GMP_jll", v"6.2.1"),
```

Comment:
This isn't a requirement, so I don't believe this does what you think it does

Reply:
I’m getting the following error in #3847:

```
ERROR: LoadError: Unsatisfiable requirements detected for package GMP_jll [781609d7]:
 GMP_jll [781609d7] log:
 ├─possible versions are: 6.1.2-6.2.1 or uninstalled
 ├─restricted to versions 6.2.1 by an explicit requirement, leaving only versions 6.2.1
 └─restricted by compatibility requirements with MPFR_jll [3a97d323] to versions: 6.2.0 — no versions left
   └─MPFR_jll [3a97d323] log:
     ├─possible versions are: 4.0.2-4.1.1 or uninstalled
     └─restricted to versions 4.1.1 by an explicit requirement, leaving only versions 4.1.1
Stacktrace:
```

My assumption was that this requirement must be coming from this `build_tarball.jl` file, is there somewhere else I should look?

Reply:
https://github.com/JuliaPackaging/Yggdrasil/pull/3847#discussion_r742415972

Reply:
Based on the BinaryBuilder.jl, I don't understand why the above line isn't a hard requirement for version GMP 6.2.0, as the docs seem to suggest that they create one: https://docs.binarybuilder.org/stable/reference/#BinaryBuilderBase.Dependency. (I'm likely missing/overlooking something.)

Would it make sense to change it to a minimum bound instead? E.g. swapping `Dependency("GMP_jll", v"6.2.0")` for `Dependency("GMP_jll"; v"6.2.0")`?

Reply:
Maybe you two mean something else with "(hard)requirement"

The second argument in `Dependency("GMP_jll", v"6.2.1")` is the `build_version`; it may is a "hard requirement" but only during build time; i.e., it only affects what is done when creating MPFR_jll, but it is not stored anywhere, in particular not in the generated JLL; so it doesn't introduce a requirement when *using* the JLL, as you are doing in PR #3847. To restrict the versions during "runtime" (i.e., changing the `compat` requirement in `Project.toml` of the generated JLL), the `compat` argument exists.

But this should *not* be done here, as GMP_jll is bundled with Julia, so you cannot (must not!) select its version freely; if you do that, the resulting JLL will not be loadable in all Julia versions (indeed, if  you were to fix it to require *exactly* GMP 6.2.1, it would only work in Julia versions using exactly that GMP version).

In general, we thus don't set a `compat` requirement for GMP_jll. As to the `build_version`, one should always use the oldest version possible (which for Julia >= 1.6 means GMP >= 6.2.0), as binaries built against a newer GMP may not load when an older GMP is present, introducing compatibility issues. On the other hand, a binary built against an older GMP will (well, "should", assuming nobody screwed up in a major way) work with a newer one.

Hope that clarifies it a bit.

Reply:
Thanks! Gotcha!

Reply:
One more question about this, it looks like GMP 6.2.1 is a bug fix release (stated here: https://gmplib.org/list-archives/gmp-announce/2020-November/000049.html, also based on semantic versioning). Might it not make sense to use 6.2.1 then everywhere we currently require 6.2.0? I get the argument for a minor or major release, but the point releases (according to semantic versioning logic, but no guarantees) should be strictly superior, right?

Reply:
We do dynamic loading of shared libraries, not static linking, as long as the libraries have the same API/ABI, it doesn't matter at all what you use when building.  On macOS systems libraries aren't event real libraries but stub files (`.tbd`) with the list of symbols provided and stop.